test_service broken with the move of replicatedservice from templates to application-dm-templates directory.

This was broken in #133 

2015/12/03 14:39:37 expandybird: expand: handling request:POST /expand
[restful] 2015/12/03 14:39:37 log.go:26: [restful] recover from panic situation: - runtime error: invalid memory address or nil pointer dereference
    /usr/local/go/src/runtime/asm_amd64.s:437
    /usr/local/go/src/runtime/panic.go:423
    /usr/local/go/src/runtime/panic.go:42
    /usr/local/go/src/runtime/sigpanic_unix.go:24
    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/expandybird/expander/expander.go:194
    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/expandybird/service/service.go:70
    /home/vaikas/projects/dmos/src/github.com/emicklei/go-restful/container.go:249
    /home/vaikas/projects/dmos/src/github.com/emicklei/go-restful/container.go:95
    /usr/local/go/src/net/http/server.go:1422
    /usr/local/go/src/net/http/server.go:1699
    /home/vaikas/projects/dmos/src/github.com/emicklei/go-restful/container.go:264
    <autogenerated>:32
    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/util/httputil.go:45
    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/expandybird/service/service_test.go:120
    /usr/local/go/src/testing/testing.go:456
    /usr/local/go/src/runtime/asm_amd64.s:1696
--- FAIL: TestServiceWrapper (0.00s)
	service_test.go:36: cannot create template for test case (cannot read import file (open ../../templates/replicatedservice/v1/replicatedservice.py: no such file or directory): ../../templates/replicatedservice/v1/replicatedservice.py): expect error for invalid HTTP verb
	service_test.go:36: cannot create template for test case (cannot read import file (open ../../templates/replicatedservice/v1/replicatedservice.py: no such file or directory): ../../templates/replicatedservice/v1/replicatedservice.py): expect error for invalid URL path
	service_test.go:36: cannot create template for test case (cannot read import file (open ../../templates/replicatedservice/v1/replicatedservice.py: no such file or directory): ../../templates/replicatedservice/v1/replicatedservice.py): expect error for invalid content type
	service_test.go:36: cannot create template for test case (cannot read import file (open ../../templates/replicatedservice/v1/replicatedservice.py: no such file or directory): ../../templates/replicatedservice/v1/replicatedservice.py): expect success
	service_test.go:128: error in test case 'expect success': test returned code:500, status: [restful] recover from panic situation: - runtime error: invalid memory address or nil pointer dereference
		    /usr/local/go/src/runtime/asm_amd64.s:437
		    /usr/local/go/src/runtime/panic.go:423
		    /usr/local/go/src/runtime/panic.go:42
		    /usr/local/go/src/runtime/sigpanic_unix.go:24
		    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/expandybird/expander/expander.go:194
		    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/expandybird/service/service.go:70
		    /home/vaikas/projects/dmos/src/github.com/emicklei/go-restful/container.go:249
		    /home/vaikas/projects/dmos/src/github.com/emicklei/go-restful/container.go:95
		    /usr/local/go/src/net/http/server.go:1422
		    /usr/local/go/src/net/http/server.go:1699
		    /home/vaikas/projects/dmos/src/github.com/emicklei/go-restful/container.go:264
		    <autogenerated>:32
		    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/util/httputil.go:45
		    /home/vaikas/projects/dmos/src/github.com/kubernetes/deployment-manager/expandybird/service/service_test.go:120
		    /usr/local/go/src/testing/testing.go:456
		    /usr/local/go/src/runtime/asm_amd64.s:1696

--- a/expandybird/expander/expander_test.go
+++ b/expandybird/expander/expander_test.go
@@ -26,7 +26,7 @@ import (
 const invalidFileName = "afilethatdoesnotexist"
 
 var importFileNames = []string{
-	"../../templates/replicatedservice/v1/replicatedservice.py",
+	"../test/replicatedservice.py",
 }
 
 var outputFileName = "../test/ExpectedOutput.yaml"

--- a/expandybird/service/service_test.go
+++ b/expandybird/service/service_test.go
@@ -66,7 +66,7 @@ const (
 )
 
 var importFileNames = []string{
-	"../../templates/replicatedservice/v1/replicatedservice.py",
+	"../test/replicatedservice.py",
 }
 
 type ServiceWrapperTestCase struct {

--- a/expandybird/test/replicatedservice.py
+++ b/expandybird/test/replicatedservice.py
@@ -1,0 +1,182 @@
+
+######################################################################
+# Copyright 2015 The Kubernetes Authors All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+######################################################################
+
+"""Defines a ReplicatedService type by creating both a Service and an RC.
+
+This module creates a typical abstraction for running a service in a
+Kubernetes cluster, namely a replication controller and a service packaged
+together into a single unit.
+"""
+
+import yaml
+
+SERVICE_TYPE_COLLECTION = 'Service'
+RC_TYPE_COLLECTION = 'ReplicationController'
+
+
+def GenerateConfig(context):
+  """Generates a Replication Controller and a matching Service.
+
+  Args:
+    context: Template context. See schema for context properties.
+
+  Returns:
+    A Container Manifest as a YAML string.
+  """
+  # YAML config that we're going to create for both RC & Service
+  config = {'resources': []}
+
+  name = context.env['name']
+  container_name = context.properties.get('container_name', name)
+  namespace = context.properties.get('namespace', 'default')
+
+  # Define things that the Service cares about
+  service_name = context.properties.get('service_name', name + '-service')
+  service_type = SERVICE_TYPE_COLLECTION
+
+  # Define things that the Replication Controller (rc) cares about
+  rc_name = name + '-rc'
+  rc_type = RC_TYPE_COLLECTION
+
+  service = {
+      'name': service_name,
+      'type': service_type,
+      'properties': {
+          'apiVersion': 'v1',
+          'kind': 'Service',
+          'metadata': {
+              'name': service_name,
+              'namespace': namespace,
+              'labels': GenerateLabels(context, service_name),
+          },
+          'spec': {
+              'ports': [GenerateServicePorts(context, container_name)],
+              'selector': GenerateLabels(context, name)
+          }
+      }
+  }
+  set_up_external_lb = context.properties.get('external_service', None)
+  if set_up_external_lb:
+    service['properties']['spec']['type'] = 'LoadBalancer'
+  config['resources'].append(service)
+
+  rc = {
+      'name': rc_name,
+      'type': rc_type,
+      'properties': {
+          'apiVersion': 'v1',
+          'kind': 'ReplicationController',
+          'metadata': {
+              'name': rc_name,
+              'namespace': namespace,
+              'labels': GenerateLabels(context, rc_name),
+          },
+          'spec': {
+              'replicas': context.properties['replicas'],
+              'selector': GenerateLabels(context, name),
+              'template': {
+                  'metadata': {
+                      'labels': GenerateLabels(context, name),
+                  },
+                  'spec': {
+                      'containers': [
+                          {
+                              'env': GenerateEnv(context),
+                              'name': container_name,
+                              'image': context.properties['image'],
+                              'ports': [
+                                  {
+                                      'name': container_name,
+                                      'containerPort': context.properties['container_port'],
+                                  }
+                              ]
+                          }
+                      ]
+                  }
+              }
+          }
+      }
+  }
+
+  config['resources'].append(rc)
+  return yaml.dump(config)
+
+
+def GenerateLabels(context, name):
+  """Generates labels either from the context.properties['labels'] or 
+     generates a default label 'app':name
+
+  We make a deep copy of the context.properties['labels'] section to avoid
+  linking in the yaml document, which I believe reduces readability of the
+  expanded template. If no labels are given, generate a default 'app':name.
+
+  Args:
+    context: Template context, which can contain the following properties:
+             labels - Labels to generate
+
+  Returns:
+    A dict containing labels in a name:value format
+  """
+  tmp_labels = context.properties.get('labels', None)
+  ret_labels = {'app': name}
+  if isinstance(tmp_labels, dict):
+    for key, value in tmp_labels.iteritems():
+      ret_labels[key] = value
+  return ret_labels
+
+
+def GenerateServicePorts(context, name):
+  """Generates a ports section for a service.
+
+  Args:
+    context: Template context, which can contain the following properties:
+             service_port - Port to use for the service
+             target_port - Target port for the service
+             protocol - Protocol to use.
+
+  Returns:
+    A dict containing a port definition
+  """
+  service_port = context.properties.get('service_port', None)
+  target_port = context.properties.get('target_port', None)
+  protocol = context.properties.get('protocol')
+
+  ports = {}
+  if name:
+    ports['name'] = name
+  if service_port:
+    ports['port'] = service_port
+  if target_port:
+    ports['targetPort'] = target_port
+  if protocol:
+    ports['protocol'] = protocol
+
+  return ports
+
+def GenerateEnv(context):
+  """Generates environmental variables for a pod.
+
+  Args:
+    context: Template context, which can contain the following properties:
+             env - Environment variables to set.
+
+  Returns:
+    A list containing env variables in dict format {name: 'name', value: 'value'}
+  """
+  env = []
+  tmp_env = context.properties.get('env', [])
+  for entry in tmp_env:
+    if isinstance(entry, dict):
+      env.append({'name': entry.get('name'), 'value': entry.get('value')})
+  return env

